pkg/validation/interfaces/validator.go

```diff
@@ -0,0 +1,92 @@
+package validator
+
+import (
+	"sync"
+
+	"github.com/operator-framework/api/pkg/validation/errors"
+)
+
+// ValidatorFunc returns a ManifestResult containing errors and warnings from
+// validating some object. These are typically closures or methods.
+type ValidatorFunc func() errors.ManifestResult
+
+// ValidationFuncs is a set of validation functions.
+type ValidatorFuncs []ValidatorFunc
+
+// Validator is an interface for validating arbitrary objects. A Validator
+// returns a set of functions that each validate some underlying object.
+// This allows the implementer to easily break down each validation step
+// into discrete units if they wish, and perform either lazy or active
+// validation.
+type Validator interface {
+	// GetFuncs takes a list of arbitrary objects and returns a set of functions
+	// that each validate some object from the list.
+	GetFuncs(...interface{}) ValidatorFuncs
+}
+
+// Validators is a set of Validator's that can be run via Apply.
+type Validators []Validator
+
+// Apply collects validator functions from each Validator in vals by collecting
+// the appropriate functions for each obj in objs, then invokes them,
+// collecting the results. Use Apply in code that:
+// - Uses more than one Validator in one call.
+// - Want active validation.
+func (vals Validators) Apply(objs ...interface{}) (results []errors.ManifestResult) {
+	if len(vals) == 0 {
+		return nil
+	}
+	if len(vals) == 1 {
+		return vals[0].GetFuncs(objs...).runP()
+	}
+	queue := make(chan []errors.ManifestResult, 1)
+	wg := &sync.WaitGroup{}
+	wg.Add(len(vals))
+	for _, val := range vals {
+		go func(v Validator) {
+			funcResults := []errors.ManifestResult{}
+			for _, result := range v.GetFuncs(objs...).runP() {
+				if result.HasError() || result.HasWarn() {
+					funcResults = append(funcResults, result)
```

Comment:
We don't know how many funcs are returned by `v.GetFuncs()` so we can't add the correct delta outside of the goroutine, i.e. we can't know the correct number of times to call `wg.Done()`.

Reply:
> We don't know how many funcs are returned by v.GetFuncs() so we can't add the correct delta outside of the goroutine

I'm going to assume that most of the work goes on when executing a validation function, rather than in getting them with `v.GetFuncs()` (correct me if I'm wrong). Using this assumption, we can synchronously get the functions, and execute them concurrently. 

I decided to add some more detail and ended up making something simpler than what I originally described. If we assume _no changes_ to the current interfaces:

```go
func (vals Validators) Apply(objs ...interface{}) (results []errors.ManifestResult) {
	if len(vals) == 0 {
		return nil
	}

        var wg sync.WaitGroup
        var mux sync.Mutex
	for _, val := range vals {
                // Assume it doesn't take a long time to get the set of functions
		for _, fn := range v.GetFuncs(objs...) {
                        // Synchronously add to the WaitGroup for each function we need to evaluate
                        wg.Add(1)

                        // Spin up a goroutine to handle the work and continue to the next function
                        go func() {
                             // Ensure we let the outside world know when we finish
                             defer wg.Done()

                             // Do the long-running work
                             result := fn()
			    if result == nil || !result.HasError() && !result.HasWarn() {
                                 return
			    }

                             // Aggregate the results synchronously
                             mux.Lock()
                             defer mux.Unlock()
                             results = append(results, result)

                             return
			}
		}
	}
        
        // Wait for all goroutines to return
	wg.Wait()

        // at this point, results is populated with all validator results
	return
}
```

Let me know if this makes sense, or maybe I'm missing something subtle.